File: Algorithm_generating_annotation_exon.py
#!/usr/bin/env python
'''
Algorithm for generating annotation information structure for each exon (Python
script to parse exons from ChrAll_knownGene.txt file)

USAGE:
    python Algorithm_generating_annotation_exon.py ChrAll_knownGene.txt
'''

import sys, os
import pandas as pd, numpy as np

def argsCheck(numArgs):
    '''Checks if in proper number of arguments are passed; gives instructions on
    proper use'''
    if len(sys.argv) < numArgs or len(sys.argv) > numArgs:
        print("\nPlease supply ONLY ONE file to parse.\n\nCorrect Syntax is:\n    python " + sys.argv[0] + " ChrAll_knownGene.txt\n\nThe output file will be the input file name with \".exons\" appended.\n")
        exit(1)  # Aborts program. (exit(1) indicates that an error occurred)

# Uses the function defined above to check if the number of arguments is correct
argsCheck(2)

#number of bases on exon boundary
exonboundary_offset=1
up_flank=2000
down_flank=500

inFile = sys.argv[1]  # Stores file one for input checking; note that sys.argv[0] is this script file name
inFilet=os.path.splitext(inFile)[1]
try:
    print("Reading your data file "+inFile+" ...")
    if inFilet=='.xls' or inFilet=='.xlsx':
        a = pd.read_excel(inFile)
    else:
        a = pd.read_csv(inFile, delimiter="\t")
except IOError:  #handling Exceptions
    print("Failed to open " + inFile)
    exit(1)

a.insert(0, 'bin', 'NOINFO')
a.rename(columns={'#name': 'name','exonStarts':'exStart','exonEnds':'exEnd'},
         inplace=True)
a['exStart'] = a['exStart'].str.rstrip(',')
a['exEnd'] = a['exEnd'].str.rstrip(',')
b=a.set_index(['bin','name','chrom','strand','txStart','txEnd','cdsStart','cdsEnd','exonCount','proteinID','alignID']).apply(lambda x: x.str.split(',').explode())
b['ExonNumber'] = b.groupby(b.index).cumcount()+1
b['ExonNumber-'] = - b.groupby(b.index).cumcount()
c=b.reset_index()
c.loc[c.strand == '-','ExonNumber'] = c.exonCount + c['ExonNumber-']
c.drop(columns=['proteinID', 'alignID', 'ExonNumber-'],inplace=True)
c=c.astype({'exStart': 'int32','exEnd': 'int32'})
c['intronStart']=c['exEnd'].shift(1) + exonboundary_offset
c['intronEnd']=c['exStart'] - exonboundary_offset
c.loc[c.txStart == c.exStart,["intronStart","intronEnd"]] = np.nan

c['utr5Start']=np.nan
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart),'utr5Start']=c.exStart
c.loc[(c.strand=='-')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exStart>c.cdsEnd),
      'utr5Start']=c.exStart
c.loc[(c.strand=='-')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exStart<=c.cdsEnd),
      'utr5Start']=c.cdsEnd + 1
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd),
      'utr5Start']=c.cdsEnd + 1

c['utr5End']=np.nan
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart)&(c.exEnd<c.cdsStart),
      'utr5End']=c.exEnd
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart),
      'utr5End']=c.cdsStart - 1
c.loc[(c.strand=='-')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd),
      'utr5End']=c.exEnd
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd),
      'utr5End']=c.exEnd

c['utr3Start']=np.nan
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart),'utr3Start']=c.exStart
c.loc[(c.strand=='+')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exStart>c.cdsEnd),
      'utr3Start']=c.exStart
c.loc[(c.strand=='+')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exStart<=c.cdsEnd),
      'utr3Start']=c.cdsEnd + 1
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd),
      'utr3Start']=c.cdsEnd + 1

c['utr3End']=np.nan
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart)&(c.exEnd<c.cdsStart),
      'utr3End']=c.exEnd
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart),
      'utr3End']=c.cdsStart - 1
c.loc[(c.strand=='+')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd),
      'utr3End']=c.exEnd
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd),
      'utr3End']=c.exEnd

c['upstreamStart']=np.nan
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart)&(c.exStart==c.txStart),
      'upstreamStart']=c.txStart - up_flank
c.loc[(c.strand=='-')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exEnd==c.txEnd),
      'upstreamStart']=c.txEnd + 1
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd)&(c.exEnd==c.txEnd),
      'upstreamStart']=c.txEnd + 1

c['upstreamEnd']=np.nan
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart)&(c.exStart==c.txStart),
      'upstreamEnd']=c.txStart - 1
c.loc[(c.strand=='-')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exEnd==c.txEnd),
      'upstreamEnd']=c.txEnd + up_flank
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd)&(c.exEnd==c.txEnd),
      'upstreamEnd']=c.txEnd + up_flank

c['downstreamStart']=np.nan
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart)&(c.exStart==c.txStart),
      'downstreamStart']=c.txStart - down_flank
c.loc[(c.strand=='+')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exEnd==c.txEnd),
      'downstreamStart']=c.txEnd + 1
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd)&(c.exEnd==c.txEnd),
      'downstreamStart']=c.txEnd + 1

c['downstreamEnd']=np.nan
c.loc[(c.strand=='-')&(c.exStart<=c.cdsStart)&(c.exStart==c.txStart),
      'downstreamEnd']=c.txStart - 1
c.loc[(c.strand=='+')&(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exEnd==c.txEnd),
      'downstreamEnd']=c.txEnd + down_flank
c.loc[(c.strand=='+')&(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd)&(c.exEnd==c.txEnd),
      'downstreamEnd']=c.txEnd + down_flank

c['cdsStart_']=np.nan
c.loc[(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exStart<=c.cdsEnd),
      'cdsStart_']=c.exStart
c.loc[(c.exStart>c.cdsStart)&(c.exEnd<=c.cdsEnd),'cdsStart_']=c.exStart
c.loc[(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart),'cdsStart_']=c.cdsStart

c['cdsEnd_']=np.nan
c.loc[(c.exStart>c.cdsStart)&(c.exEnd>c.cdsEnd)&(c.exStart<=c.cdsEnd),
      'cdsEnd_']=c.cdsEnd
c.loc[(c.exStart>c.cdsStart)&(c.exEnd<=c.cdsEnd),'cdsEnd_']=c.exEnd
c.loc[(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd<c.cdsEnd),
      'cdsEnd_']=c.exEnd
c.loc[(c.exStart<=c.cdsStart)&(c.exEnd>=c.cdsStart)&(c.exEnd>=c.cdsEnd),
      'cdsEnd_']=c.cdsEnd

c.drop(columns=['cdsStart', 'cdsEnd', 'strand', 'exonCount'],inplace=True)
c.rename(columns={'cdsStart_': 'cdsStart','cdsEnd_':'cdsEnd'},
         inplace=True)
c=c[['bin','name','chrom','txStart','txEnd','cdsStart','cdsEnd','ExonNumber',
     'exStart','exEnd','intronStart','intronEnd','utr5Start','utr5End','utr3Start',
     'utr3End','upstreamStart','upstreamEnd','downstreamStart','downstreamEnd']]
c=c.astype({'cdsStart': 'Int32','cdsEnd': 'Int32','intronStart': 'Int32',
            'intronEnd': 'Int32','utr5Start': 'Int32','utr5End': 'Int32',
            'utr3Start': 'Int32','utr3End': 'Int32','upstreamStart': 'Int32',
            'upstreamEnd': 'Int32','downstreamStart': 'Int32','downstreamEnd':'Int32',})

outFile = inFile + ".exons"
try:
    print("Saving data to file "+outFile+" ...")
    c.to_csv(outFile,index=False,sep='\t',na_rep='NA')
except IOError:
    print("Failed to create " + outFile)
    exit(1)

File: Algorithm_generating_annotation_exon.pl
#!/usr/bin/perl

# Algorithm for generating annotation information structure for each exon (perl script to parse exons from ChrAll_knownGene.txt file)
# USAGE: perl Algorithm_generating_annotation_exon.perl ChrAll_knownGene.txt

if(! @ARGV)
{
  die  "Correct Syntax is: Algorithm_generating_annotation_exon.perl <fileneame> \n\nPlease supply a file to parse.  The output file will be the input file name with \".exons\" appended\n\n";
}

$infile = shift @ARGV;

$outfile = $infile . ".exons";

#number of bases on exon boundary
$exonboundary_offset=1;
$up_flank=2000;
$down_flank=500;

open INFILE, "<$infile";
open OUTFILE, ">$outfile";
print OUTFILE "bin\tname\tchrom\ttxStart\ttxEnd\tcdsStart\tcdsEnd\tExonNumber\texStart\texEnd\tintronStart\tintronEnd\tUTR5Start\tUTR5End\tUTR3Start\tUTR3End\tupstreamStart\tupstreamEnd\tdownstreamStart\tdownstreamEnd\n";

while(<INFILE>)  #use the Perl while loop to read a file line by line to the end of the file
{
  chomp;  # avoid \n on last field
  if($_ =~ /^\#/)  #boolean condition; \# if u do want # character, you'll have to quote it with a backslash
  {
    next;  #The next command starts the next iteration of the loop
  }

   ($name, $chrom, $strand, $txStart, $txEnd, $cdsStart,
  $cdsEnd, $exonCount, $exonStarts, $exonEnds, $proteinID, $alignID)=split(/\t/,$_,12);

  @exstarts=split(/,/,$exonStarts);
  @exends=split(/,/,$exonEnds);

  my $bin = "NOINFO";  #for knownGene table parse only!!

  my $CDSStart = "NA";
  my $CDSEnd = "NA";
  my $intronStart = "NA";
  my $intronEnd = "NA";
  my $UTR5Start = "NA";
  my $UTR5End = "NA";
  my $UTR3Start = "NA";
  my $UTR3End = "NA";
  my $upstream = "NA";
  my $downstream = "NA";


  $count=0;
  while($count < $exonCount)
  {
    #if this is the first exon
    if($txStart == $exstarts[$count])
    {
      $intronStart = "NA";
      $intronEnd = "NA";
    }
    else #assign intron to the next Exon
    {
      $intronStart = $exends[$count - 1] + $exonboundary_offset;
      $intronEnd = $exstarts[$count] - $exonboundary_offset;
    }
    #if cdsStart is located in this exon
    if($strand eq "+")
    {
      $ex_num = $count + 1;
      if($exstarts[$count] > $cdsStart) #1
      {
        if($exends[$count] > $cdsEnd) #1.1
        {
          if($exstarts[$count] > $cdsEnd) #1.1.1
          {
            $CDSStart = "NA";
            $CDSEnd = "NA";
            $UTR3Start = $exstarts[$count];
          }
          else#1.1.2
          {
            $CDSStart = $exstarts[$count];
            $CDSEnd = $cdsEnd;
            $UTR3Start = $cdsEnd + 1;
          }
          $UTR3End = $exends[$count];
          if($exends[$count] == $txEnd) #if this is the last exon
          {
            $downstreamStart = $txEnd + 1;
            $downstreamEnd = $txEnd + $down_flank;
          }
          else
          {
            $downstreamStart = "NA";
            $downstreamEnd = "NA";
          }
        }
        else#1.2
        {
          $CDSStart = $exstarts[$count];
          $CDSEnd = $exends[$count];
          $UTR3Start = "NA";
          $UTR3End = "NA";
          $downstreamStart = "NA";
          $downstreamEnd = "NA";
        }
        $UTR5Start = "NA";
        $UTR5End = "NA";
        $upstreamStart = "NA";
        $upstreamEnd = "NA";
      }
      else#2
      {
        if($exends[$count] < $cdsStart) #2.1
        {
          $CDSStart = "NA";
          $CDSEnd = "NA";
          $UTR3Start = "NA";
          $UTR3End = "NA";
          $UTR5End = $exends[$count];
          $downstreamStart = "NA";
          $downstreamEnd = "NA";
        }
        else #2.2
        {
          $CDSStart = $cdsStart;
          $UTR5End = $cdsStart - 1;
          if($exends[$count] < $cdsEnd) #2.2.1
          {
            $CDSEnd = $exends[$count];
            $UTR3Start = "NA";
            $UTR3End = "NA";
            $downstreamStart = "NA";
            $downstreamEnd = "NA";
          }
          else #2.2.2
          {
            $CDSEnd = $cdsEnd;
            $UTR3Start = $cdsEnd + 1;
            $UTR3End = $exends[$count];
            if($exends[$count] == $txEnd) #if this is the last exon
            {
              $downstreamStart = $txEnd + 1;
              $downstreamEnd = $txEnd + $down_flank;
            }
            else
            {
              $downstreamStart = "NA";
              $downstreamEnd = "NA";
            }
          }
        }
        $UTR5Start = $exstarts[$count];
        if($exstarts[$count] == $txStart)#if this is the first exon
        {
          $upstreamStart = $txStart - $up_flank;
          $upstreamEnd = $txStart - 1;
        }
        else
        {
          $upstreamStart = "NA";
          $upstreamEnd = "NA";
        }
      }
    }
    else #negative strand
    {
      $ex_num = $exonCount - $count;
      if($exstarts[$count] > $cdsStart)#1
      {
        if($exends[$count] > $cdsEnd)#1.1
        {
          if($exstarts[$count] > $cdsEnd) #1.1.1
          {
            $CDSStart = "NA";
            $CDSEnd = "NA";
            $UTR5Start = $exstarts[$count];
          }
          else#1.1.2
          {
            $CDSStart = $exstarts[$count];
            $CDSEnd = $cdsEnd;
            $UTR5Start = $cdsEnd + 1;
          }
          $UTR5End = $exends[$count];
          if($exends[$count] == $txEnd) #if this is the last exon
          {
            $upstreamStart = $txEnd + 1;
            $upstreamEnd = $txEnd + $up_flank;
          }
          else
          {
            $upstreamStart = "NA";
            $upstreamEnd = "NA";
          }
        }
        else#1.2
        {
          $CDSStart = $exstarts[$count];
          $CDSEnd = $exends[$count];
          $UTR5Start = "NA";
          $UTR5End = "NA";
          $upstreamStart = "NA";
          $upstreamEnd = "NA";
        }
        $UTR3Start = "NA";
        $UTR3End = "NA";
        $downstreamStart = "NA";
        $downstreamEnd = "NA";
      }
      else#2
      {
#if($name eq "uc009ygf.2")
#{
#  print $exstarts[$count], "\t", $txStart, "\n";
#}
        if($exends[$count] < $cdsStart)#2.1
        {
          $CDSStart = "NA";
          $CDSEnd = "NA";
          $UTR3End = $exends[$count];
          $UTR5Start = "NA";
          $UTR5End = "NA";
          $upstreamStart = "NA";
          $upstreamEnd = "NA";
        }
        else#2.2
        {
          if($exends[$count] < $cdsEnd)#2.2.1
          {
            $CDSEnd = $exends[$count];
            $UTR5Start = "NA";
            $UTR5End = "NA";
            $upstreamStart = "NA";
            $upstreamEnd = "NA";
          }
          else#2.2.2
          {
            $CDSEnd = $cdsEnd;
            $UTR5Start = $cdsEnd + 1;
            $UTR5End = $exends[$count];
            if($exends[$count] == $txEnd) #if this is the last exon
            {
              $upstreamStart = $txEnd + 1;
              $upstreamEnd = $txEnd + $up_flank;
            }
            else
            {
              $upstreamStart = "NA";
              $upstreamEnd = "NA";
            }
          }
          $CDSStart = $cdsStart;
          $UTR3End = $cdsStart - 1;
        }
        $UTR3Start = $exstarts[$count];
        if($exstarts[$count] == $txStart) #if this is the first exon
        {
          $downstreamStart = $txStart - $down_flank;
          $downstreamEnd = $txStart - 1;
        }
        else
        {
          $downstreamStart = "NA";
          $downstreamEnd = "NA";
        }
      }
    }
    print OUTFILE "$bin\t$name\t$chrom\t$txStart\t$txEnd\t$CDSStart\t$CDSEnd\t$ex_num\t$exstarts[$count]\t$exends[$count]\t$intronStart\t$intronEnd\t$UTR5Start\t$UTR5End\t$UTR3Start\t$UTR3End\t$upstreamStart\t$upstreamEnd\t$downstreamStart\t$downstreamEnd\n";
    $count++;
  }##inside while
}###outside while
